docs/ref/csp.txt

```diff
@@ -0,0 +1,296 @@
+   :synopsis: Middleware for Content Security Policy headers
+
+Content Security Policy (CSP) is a web security standard that helps prevent
+attacks by controlling the sources from which content can be loaded. CSP
+enforcement plays an important role in a comprehensive :doc:`security strategy
+</topics/security>`.
+
+.. _using-csp:
+
+Using CSP
+=========
+
+To enable CSP in your Django project:
+
+1. Add the CSP middleware to your ``MIDDLEWARE`` setting::
+
+      MIDDLEWARE = [
+          # ...
+          "django.middleware.csp.ContentSecurityPolicyMiddleware",
+          # ...
+      ]
+
+2. Configure your CSP policies in your ``settings.py`` using either
+   ``SECURE_CSP`` or ``SECURE_CSP_REPORT_ONLY``::
+
+      from django.middleware import csp
+
+      SECURE_CSP = {
+          "DIRECTIVES": {
+              "default-src": [csp.SELF],
+              "script-src": [csp.SELF],
+              "style-src": [csp.SELF],
+              # Add more directives as needed
+          }
+      }
+
+      # Or for report-only mode:
+      SECURE_CSP_REPORT_ONLY = {
+          "DIRECTIVES": {
+              "default-src": [csp.SELF],
+              "script-src": [csp.SELF],
+              "style-src": [csp.SELF],
+              # Add more directives as needed
+          }
+      }
+
+3. To use CSP nonces for inline scripts or styles, include ``csp.NONCE`` in your
+   directives::
+
+      SECURE_CSP = {
+          "DIRECTIVES": {
+              "script-src": [csp.SELF, csp.NONCE],
+              # Other directives...
+          }
+      }
+
+Nonce Usage
+===========
+
+To use nonces in your CSP policy:
+
+1. Add the CSP context processor to your ``TEMPLATES`` setting::
+
+      TEMPLATES = [
+          {
+              "BACKEND": "django.template.backends.django.DjangoTemplates",
+              "OPTIONS": {
+                  "context_processors": [
+                      # ...
+                      "django.template.context_processors.csp_nonce",
+                  ],
+              },
+          },
+      ]
+
+This context processor makes the CSP nonce available in all templates.
+
+2. In your templates, use the ``request.csp_nonce``:
+
+   .. code-block:: html
+
+      <script nonce="{{ request.csp_nonce }}">
+          // Your inline JavaScript
+      </script>
+
+The CSP middleware will automatically generate and include a nonce in the CSP
+header for each request.
+
+.. note::
+
+    The CSP nonce is included in the header only if ``request.csp_nonce`` is
+    accessed at some point between when the middleware processes the request and
+    the response. This ensures the nonce is generated and included only when
+    necessary.
+
+    If the Django CSP settings do not explicitly include the ``csp.NONCE``
+    directive and ``request.csp_nonce`` is accessed in a template, the nonce
+    will appear in the content but not in the CSP header. This discrepancy may
+    lead to confusion or unexpected behavior.
+
+    To avoid such issues:
```

Comment:
Maybe when debug is True there should be some way to check this (check for the csp.NONCE in the settings if request.csp_nonce is accessed, that might still mismatch between where csp_nonce is used and where it is defined in the header, but it prevents the obvious mistake of not including it in any header.

Silently failing will just lead to lot of questions on discord about why doesn't this work.